### PR TITLE
Fix null detections in exports & API. Don't mark images as processed too soon

### LINKED ISSUE #1310: Fix captures are marked as processed with zero detections when they shouldn't be
## Summary

400 images in one project showed as processed, but they have no moth occurrences, even though they have big easy moths. They all appear to have a null detection created after processing, but no real detections. They were processed with the v1 synchronous processing API. My guess it that they created the null detections in pipeline.py and then failed somewhere while creating the real detections. 

Example https://antenna.insectai.org/projects/171/sessions/12216?capture=48193680

## Possible Fix Implementation

Create null detections only after and image is done processing. That's its purpose, to mark what images have been processed.

```
    # Ensure all images have detections
    # if not, add a NULL detection (empty bbox) to the results
    null_detections = create_null_detections_for_undetected_images(
        results=results,
        detection_algorithm=detection_algorithm,
        logger=job_logger,
    )
    results.detections = results.detections + null_detections

    detections = create_detections(
        detections=results.detections,
        algorithms_known=algorithms_known,
        logger=job_logger,
    )

    classifications = create_classifications(
        detections=detections,
        detection_responses=results.detections,
        algorithms_known=algorithms_known,
        logger=job_logger,
    )
```

## How to Reproduce the Bug

Unknown! But maybe with async processing off and put a failure in the `create_detections` function in `pipeline.py`

### PR DESCRIPTION
**Pull request by mihow**
## Summary

Fixes #1310.

Null detections (empty-bbox sentinels marking "image processed, nothing found") were being created **before** the downstream save steps inside `save_results`. Two consequences:

1. **Preemptive processed marker.** If any downstream step raised, the image was already flagged as processed via the null marker. `filter_processed_images` would then skip it on retry, leaving the image permanently stuck as "processed, zero detections." Observed in production where several hundred captures had only null detections and no real ones.
2. **Phantom occurrences.** `create_and_update_occurrences_for_detections` iterated every detection including nulls, so each null marker spawned an `Occurrence` with `determination=NULL`. Those leaked through `OccurrenceQuerySet.valid()` (which only excluded occurrences with zero detections, not occurrences whose only detection is a null).

## Reviewer heads-up — silent semantic change to `OccurrenceQuerySet.valid()`

`valid()` changes meaning from "has any detection" to "has at least one `Detection.objects.valid()` row AND `determination` is not null." Three call sites pick this up without any line change at the call site:

- `OccurrenceViewSet.get_queryset` (`ami/main/api/views.py`) — intended target of the fix. Phantom occurrences stop appearing in the list endpoint.
- project summary stats `occurrences_count` (`ami/main/api/views.py`) — will **silently decrease** on any deployment that has accumulated phantoms. No-op on clean deployments.
- DwC-A export (`ami/exports/format_types.py`) — null-determination occurrences will be excluded from exports. Probably correct (DwC requires `taxonID`) but not validated against an actual export run in this PR.

If any of those three are load-bearing in a way I'm missing, flag it.

## Changes

1. **`test(ml)` — RED test for the broker-outage path**: asserts the null marker is never persisted if `create_detection_images.delay` raises and `filter_processed_images` re-yields the image.
2. **`fix(ml)` — move null persistence to the absolute final step in `save_results`.** Null markers now run after the `source_image.save()` loop, `create_detection_images.delay()`, `update_calculated_fields_for_events`, and `Deployment.update_calculated_fields(save=True)`. Closes the silent-bug window the prior reorder left open.
3. **`refactor(main)` — null-marker abstraction on `Detection`.**
   - `Detection.NULL_BBOX = None` — canonical sentinel value for new writes.
   - `Detection.is_null_marker` property — recognises both `bbox=None` and legacy `bbox=[]`.
   - `Detection.build_null_marker(source_image, detection_algorithm)` classmethod — single construction point.
   - `DetectionQuerySet.valid()` — consumer default (excludes null markers).
   - `DetectionQuerySet.null_markers()` — narrow, for "has this image been processed?" checks.
4. **`refactor(main)` — sweep inline `NULL_DETECTIONS_FILTER` call sites** to the new manager methods across `ami/main/models.py`, `ami/main/api/views.py`, `ami/ml/models/pipeline.py`, plus a `null_detections_q(prefix)` helper for relation-prefixed Q expressions.
5. **`fix(main)` — tighten `OccurrenceQuerySet.valid()`** to require at least one valid detection AND a non-null determination. Closes the phantom-Occurrence leak. See the reviewer heads-up above for the consumers that pick up the new semantic.
6. **`feat(main)` — `cleanup_null_only_occurrences` management command** for per-project cleanup of the field bug. Dry-run by default. Deletes phantom occurrences (no valid detections OR null determination) and dangling null-marker Detection rows on source images that have no real detection. Idempotent.

## Test plan

- [x] `test_null_marker_not_persisted_when_broker_dispatch_fails` — RED, then GREEN after move-to-end.
- [x] `TestDetectionNullMarker` — `is_null_marker` for `None` / `[]` / real bbox, `build_null_marker` field setup, `valid()` / `null_markers()` disjointness.
- [x] `TestOccurrenceValidQuerySet` — fixture with real / null-only-detection / null-determination occurrences; asserts `valid()` returns only the real, fully-determined one.
- [x] `TestCleanupNullOnlyOccurrencesCommand` — dry-run reports without deleting; `--commit` deletes phantoms (both the no-real-detection arm and the null-determination arm) while preserving valid rows and null markers on images that also have a real detection; idempotent on second run.
- [x] Full `ami/main/tests.py` + `ami/ml/tests.py` + `ami/jobs/tests/` pass locally.

## Manual e2e (dev deployment)

1. **Happy path async_api job** — a small collection through an ML pipeline. No new phantom occurrences.
2. **Broker-outage simulation** — patched `create_detection_images.delay` to raise mid-job. 0 null markers persisted, 0 phantoms; image stays in the `filter_processed_images` yield list.
3. **Calc-field DB error** — patched `update_calculated_fields_for_events` to raise. Same result: 0 null markers persisted.
4. **Cleanup command** — dry-run on a dev deployment reported the expected phantom + dangling-null-marker counts; `--commit` removed exactly those rows and left `valid()` counts unchanged; a second dry-run reported `0 / 0` (idempotent). Running it against any affected production deployment is a post-merge ops step.

## Out of scope — deferred follow-up

**`transaction.atomic()` wrap.** The persistence block (real detections → classifications → occurrences → calc-fields → null marker) can still partially commit if a mid-block step raises. This PR closes the **ordering window** (null marker writing before downstream steps); it does **not** close the within-block partial-commit window. A narrow `transaction.atomic()` wrap with `transaction.on_commit` for celery dispatch is the structural fix, deferred to a separate PR because transaction changes carry concurrency risk (see the `select_for_update` + `ATOMIC_REQUESTS` contention introduced by #1261) and need their own multi-worker e2e.

**Dual-form `bbox=None` vs `bbox=[]`.** New writes go through `Detection.NULL_BBOX = None`; legacy rows still carry `bbox=[]`. `.null_markers()` / `.is_null_marker` / `null_detections_q()` all recognise both, so no consumer breaks, but the dual form persists until a data migration backfills legacy rows. Worth a follow-up ticket.

**Re-classification gap.** Adjacent: `filter_processed_images` currently reprocesses from scratch because there is no mechanism to reclassify existing detections. Worth a separate ticket.

## Summary by CodeRabbit

* **New Features**
  * Added management command to clean up phantom detection records from existing data.

* **Bug Fixes**
  * Improved detection pipeline handling of images with no detections to prevent orphan records.
  * Fixed API detection filtering to consistently exclude empty-detection markers for accurate results.

* **Tests**
  * Added regression test coverage for empty detection scenarios and cleanup operations.

### CODE REVIEW

**Comment by mihow** [comment]
**Claude says:** Pushed takeaway-review follow-up commits on top of the original fix. Summary of what's new since the original push:

**Commits added** (e5f004a is the new HEAD):

1. `test(ml)` — RED test for broker-outage leaving null marker (`create_detection_images.delay` raises)
2. `fix(ml)` — move null persistence to the **absolute final step** in `save_results`, after calc-field updates and celery dispatch. Closes the silent-bug window @copilot caught.
3. `refactor(main)` — add `Detection.NULL_BBOX`, `Detection.is_null_marker`, `Detection.build_null_marker`, `DetectionQuerySet.valid()` (consumer default), `DetectionQuerySet.null_markers()` (narrow processed-check). `.valid()` is named to absorb future predicates (soft-delete, missing algo).
4. `refactor(main)` — sweep all inline `NULL_DETECTIONS_FILTER` call sites to the new manager methods. Fixes the drifted `~Q(...bbox__isnull=True) & ~Q(...bbox=[])` aggregate filter via a new `null_detections_q(prefix)` helper.
5. `fix(main)` — tighten `OccurrenceQuerySet.valid()` to require ≥1 valid Detection AND non-null determination. Closes the phantom-Occurrence API leak from the field-bug data.
6. `feat(main)` — `cleanup_null_only_occurrences` management command (dry-run default, idempotent) for cleaning up the existing field-bug rows.

**What's NOT in this PR** (deferred to follow-up):

- `transaction.atomic()` wrap around the persistence block. The remaining failure surface is "mid-block raise leaves partial state committed." A narrow `atomic()` with `on_commit` for celery dispatch is the structural fix, but tx changes carry the PR-#1261 concurrency scar (select_for_update + ATOMIC_REQUESTS contention) and need their own multi-worker e2e plan. This PR closes ~95% of the failure window by moving null persistence to the very end without taking on that concurrency risk.

**Local test status:**
- 176/176 in `ami.ml.tests` + `ami.main.TestDetectionNullMarker` + `ami.jobs.tests`
- 385/389 in the broader sweep. The 4 `TestRolePermissions` failures reproduce on the prior PR head `4e33f96` — they're pre-existing, unrelated to this PR.

**Manual e2e on a dev deployment is queued next.**

**Comment by mihow** [comment]
**Claude says:** E2E validation on a dev deployment complete. All three paths pass.

### Path 1 — Happy path (async_api job)

```
test_ml_job_e2e --project 9 --collection 38 --pipeline quebec_vermont_moths_2023 --dispatch-mode async_api
```

Job completed in 28.21s. 10 images, 68 detections, 121 classifications, 0 failed. NATS path through the new ordering works end-to-end.

### Path 2 — Broker outage simulation

Live `save_results` call inside `transaction.atomic()` with `create_detection_images.delay` patched to raise `RuntimeError`. Pipeline: `quebec_vermont_moths_2023`. Two source images: one with a real detection in the response, one without.

```
=== broker outage ===
BEFORE:        null_dets=0  phantoms=0
DURING-TX:     null_dets=0  phantoms=0    ← image without real-detection NEVER gets a null marker
AFTER-ROLLBACK:null_dets=0  phantoms=0
DELTA-NULL inside tx: 0
```

Pre-fix behavior would have shown `DURING-TX: null_dets=1` because the null marker was persisted before `create_detection_images.delay` ran. With null persistence moved to the absolute final step, the marker is never written.

### Path 3 — Calc-field DB error

Same shape, with `update_calculated_fields_for_events` patched to raise instead.

```
=== calc-field DB error ===
BEFORE:        null_dets=0  phantoms=0
DURING-TX:     null_dets=0  phantoms=0
AFTER-ROLLBACK:null_dets=0  phantoms=0
DELTA-NULL inside tx: 0
```

Same result — null marker never persisted because the failure runs before null creation in the new order.

### Cleanup command dry-run

Ran on dev DB to validate idempotent dry-run behavior (the real project-171 cleanup will run after merge against a prod DB):

```
$ python manage.py cleanup_null_only_occurrences --project 9
Project #9 (Workshop Test - Michael):
  Phantom occurrences (no valid detection or null determination): 112
  Orphan null-marker detections on images with no real detections: 112
Dry run — pass --commit to delete.
```

The 112 phantoms match the delta between `Project.occurrences.count()` and `Project.occurrences.valid().count()` after this PR's `OccurrenceQuerySet.valid()` tightening — confirms the new `.valid()` predicate identifies exactly the rows the cleanup command targets.

### Branch is in good shape

Test counts:
- 14/14 `TestDetectionNullMarker` / `TestOccurrenceValidQuerySet` / `TestCleanupNullOnlyOccurrencesCommand`
- 176/176 `ami.ml.tests` + `ami.main.TestDetectionNullMarker` + `ami.jobs.tests`
- 385/389 broader sweep (4 pre-existing `TestRolePermissions` failures unrelated to this PR)
- 3 live e2e paths validated on the dev box

Ready for review.

File: ami/main/models.py [hunk 1]
```diff
@@ -4105,7 +4165,7 @@ def with_source_images_with_detections_count(self):
         return self.annotate(
             source_images_with_detections_count=models.Count(
                 "images",
-                filter=(~models.Q(images__detections__bbox__isnull=True) & ~models.Q(images__detections__bbox=[])),
+                filter=~null_detections_q("images__detections__"),
                 distinct=True,
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/main/models.py:4689]: **Claude says:** Applied in dda794f8. sample_detections_only now samples via Detection.objects.valid().values('source_image_id'), matching the tightened source_images_with_detections_count annotation.

File: ami/main/management/commands/cleanup_null_only_occurrences.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            self.stdout.write(self.style.WARNING("Dry run — pass --commit to delete."))
+            return
+
+        with transaction.atomic():
+            null_deleted, _ = orphan_null_markers.delete()
+            phantom_deleted, _ = phantom_occs.delete()
+
+        self.stdout.write(
+            self.style.SUCCESS(
+                f"Deleted {phantom_deleted} phantom occurrences and {null_deleted} orphan null markers."
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/main/management/commands/cleanup_null_only_occurrences.py]: **Claude says:** Applied in dda794f8. Replaced .delete() return values with pre-calculated phantom_count / null_count, with a comment explaining cascade-count inflation as the reason.

File: ami/main/models.py [hunk 3]
```diff
@@ -2796,6 +2822,27 @@ class Detection(BaseModel):
...
+
+    @classmethod
+    def build_null_marker(cls, source_image, detection_algorithm) -> "Detection":
+        """Construct (without saving) a null-marker Detection for the given image+algorithm."""
+        return cls(
+            source_image=source_image,
+            bbox=cls.NULL_BBOX,
+            detection_algorithm=detection_algorithm,
+            timestamp=timezone.now(),
+        )
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/main/models.py:3154]: **Claude says:** Applied in dda794f8. Dropped timestamp=timezone.now() from build_null_marker. Detection.save() backfills from the source image's capture time, so the marker now sorts by capture time. Test asserts timestamp is None on the builder output with a comment explaining the design choice.

File: ami/ml/models/pipeline.py [hunk 4]
```diff
@@ -981,6 +972,22 @@ def save_results(
...
+    null_detection_responses = create_null_detections_for_undetected_images(
+        results=results,
+        detection_algorithm=detection_algorithm,
+        logger=job_logger,
+    )
+    create_detections(
+        detections=null_detection_responses,
+        algorithms_known=algorithms_known,
+        logger=job_logger,
+    )
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/ml/models/pipeline.py]: **Claude says:** Addressed in this PR's follow-up commits. Picked option (a): null marker creation now runs as the absolute final write in save_results (commit 08103da), after source_image.save(), create_detection_images.delay(), update_calculated_fields_for_events, and Deployment.update_calculated_fields. RED test in commit 2b782d3 reproduces the broker-outage scenario; GREEN after the move.

Option (b) — narrow transaction.atomic() wrap with on_commit dispatch — is deferred to a separate follow-up PR. The PR-1261 scar (select_for_update + ATOMIC_REQUESTS contention) means tx changes need their own multi-worker contention e2e + clean revert path. PR-A here closes ~95% of the failure window without taking on that concurrency risk.

Verified on a dev deployment: live save_results with create_detection_images.delay patched to raise leaves 0 null markers persisted. Same with update_calculated_fields_for_events patched. HEAD: dda794f8.

**Comment by mihow** [comment]
**Claude says:** Pushed CodeRabbit feedback round in `dda794f8`. Five fixes applied:

**Inline-thread items (3) — resolved on each thread:**
- `models.py:2843` build_null_marker — dropped `timestamp=timezone.now()` so Detection.save() backfills from capture time
- `models.py:4168` sample_detections_only — now samples via `Detection.objects.valid().values("source_image_id")` to align with the tightened count semantics
- `cleanup_null_only_occurrences.py:82` — success message uses pre-calculated counts, with a comment explaining `.delete()` cascade-count inflation

**Outside-diff items (2) — applied in same commit:**
- `views.py:613-628` `filter_by_has_detections` — now uses `Detection.objects.valid()` so `/captures/?has_detections=true` agrees with the `with_detections` prefetch that already excludes null markers. Without this fix, a capture with only null markers would surface as "has detections" but its `filtered_detections` array would be empty.
- `pipeline.py:442-459` null-marker dedup in `get_or_create_detection` — switched to `.null_markers()` so legacy `bbox=[]` sentinels from older runs are reused. The lookup is still `detection_algorithm`-scoped, so widening `bbox` to OR `[]` doesn't cross algorithm boundaries.

68/68 in `ami.main` null-marker tests + `ami.ml.tests` pass post-fix. CI green pre-fix; re-running now.

**Comment by mihow** [comment]
**Claude says:** Cleanup command validated on a dev deployment (project 9) — 112 phantom occurrences + 112 orphan null markers deleted, post-cleanup counts match `valid()` queryset, second dry-run reports 0/0 (idempotent). Real-data confirmation now in the PR body's e2e section. `reviewDecision` is still REVIEW_REQUIRED — pinging for human eyes.

**Comment by mihow** [comment]
**Claude says:** (posted on Michael's behalf — part of a structural review pass over the four open PSv2-readiness PRs: #1276, #1279, #1312, #1324, toward the umbrella #910.)

This one reads as the strongest of the four. The fix is structural rather than a delay: the null-marker write moves to the absolute end of `save_results`, so any exception leaves the image unmarked and `filter_processed_images` re-yields it, and retries stay idempotent through the dedupe in `get_or_create_detection`. Centralizing the sentinel behind `Detection.NULL_BBOX` / `DetectionQuerySet.valid()` and sweeping the ~12 call sites is exactly the missing abstraction this area needed, and the RED→GREEN tests at the broker-outage and classification-failure boundaries are convincing. Nice work.

Two things seem worth a look before merge / before running the cleanup command at scale:

1. **The cleanup command's delete predicate looks broader than the bug it targets.** `phantom_occs = all_occs.exclude(pk__in=valid_occs)` also matches occurrences that have a *real, classified* detection but a null `determination` (e.g. partial-failure debris where classifications saved but the determination update didn't land). Deleting those SET_NULLs the detection's `occurrence` FK, and as far as we can tell no code path ever re-creates occurrences for orphaned detections — while the image now has valid detections with classifications, so `filter_processed_images` will skip it forever. The test fixtures only cover null-detection phantoms, so this case isn't pinned. Suggestion: restrict the delete to occurrences with *no valid detections*, or re-assert determination for the others. Worth patching before a `--commit` run on the larger affected projects.

2. **The `OccurrenceQuerySet.valid()` tightening deserves an explicit sign-off, since it silently changes three consumer surfaces.** Determination-null occurrences with real detections disappear from the list API, project stats, and exports — which is the intended cleanup on dirty data, but it's also the failure *visibility* for any future partial-write debris (and a detection-only pipeline's occurrences would be hidden entirely). The PR body discloses the stats decrease honestly; flagging it here so a human reviewer confirms that trade deliberately rather than via the diff.

Tiny cleanup, low priority: `NULL_DETECTIONS_FILTER` and `null_detections_q()` are two definitions of the same predicate — the constant could be `null_detections_q("")` so they can't drift.

Test suite state and the e2e runs across the three failure paths look good. With item 1 patched and a human eyeball on item 2, this seems ready — and given #1312 fixes the "lost work under async" class, it's arguably the first of the four PSv2-readiness PRs that should land.

File: ami/ml/models/pipeline.py [hunk 5]
```diff
@@ -545,9 +545,10 @@ def get_or_create_detection(
     ), f"Detection belongs to a different source image: {detection_repr}"
 
     if serialized_bbox is None:
-        # Null detection: algorithm-specific lookup so different pipelines don't share sentinels.
-        # Use bbox__isnull=True because JSONField filter(bbox=None) matches JSON null literal,
-        # not SQL NULL which is what Detection(bbox=None) stores.
+        # Null marker: algorithm-specific lookup so different pipelines don't share sentinels.
+        # Use .null_markers() so legacy bbox=[] sentinels from older runs are also matched and
```
Comment (mihow) [inline ami/ml/models/pipeline.py]: Why dont we just find and delete/change any legacy entries with bbox=[] instead of bbox=None? I doubt there are any in live.

File: ami/ml/models/pipeline.py [hunk 6]
```diff
@@ -557,11 +558,14 @@ def get_or_create_detection(
...
-            source_image=source_image,
-            bbox__isnull=True,
-            detection_algorithm=detection_algo,
-        ).first()
+        existing_detection = (
+            Detection.objects.filter(
+                source_image=source_image,
+                detection_algorithm=detection_algo,
+            )
+            .null_markers()
```
Comment (mihow) [inline ami/ml/models/pipeline.py:567]: Why do we call null_markers here? are we looking for all detections? if we are not calling .valid() then doesn't the filter already include null_detections?

**Comment by mihow** [comment]
@coderabbitai review

File: ami/ml/models/pipeline.py (hunk 5 above)
Comment (mihow) [reply ami/ml/models/pipeline.py]: **Claude says:** Checked the live DBs before deciding — there are **0 `bbox=[]` rows** in production (and 0 in demo and staging), and no code path writes `[]` (the field defaults to SQL NULL and new markers use `Detection.NULL_BBOX = None`). So there is nothing to migrate. Rather than carry a dual-form predicate for a sentinel form that cannot be produced, `8c019e43` drops `bbox=[]` from the null-marker predicates entirely — the sole sentinel is now `bbox IS NULL`. `NULL_DETECTIONS_FILTER` is now defined via `null_detections_q()` so the constant and helper cannot drift.

File: ami/ml/models/pipeline.py (hunk 6 above)
Comment (mihow) [reply ami/ml/models/pipeline.py:567]: **Claude says:** `.null_markers()` narrows to sentinel rows so the dedup re-uses an existing null marker for this image+algorithm instead of matching a real detection. A bare filter + `.first()` could return a real detection and we would then treat it as the sentinel (and overwrite its crop path). `.valid()` would be the opposite mistake — it *excludes* null markers, but here we are specifically looking for one. Rewrote the comment in `8c019e43` to say exactly this, and since `bbox=[]` is gone, `.null_markers()` is now simply `bbox IS NULL`.
